### DISCUSSION TITLE
Why did the call to an endpoint fail (SocketException / iOS)?

### DISCUSSION THREAD

**Discussion by fleeser** [serverpod/serverpod#631] [Q&A] [answered]
Device: iOS Simulator on macOS

When trying to make a call to my AuthEndpoint from within my App, I get the following Error:

```
flutter: Failed call: auth.signIn
flutter: SocketException: Connection refused (OS Error: Connection refused, errno = 61), address = localhost, port = 55075
```

The call looks like this:
```
AuthSignInResponse response = await client.auth.signIn(email, password);
```

main function:

```
client = Client('http://localhost:8080/')
    ..connectivityMonitor = FlutterConnectivityMonitor();
```

Since I want to implement my very own authentication flow, I did not make use of `SessionManager` and also did not pass an `AuthenticationKeyManager` to the `Client` constructor. I am manually handling when to save or retrieve which key inside my AuthRepository class, e.g.

```
Future<void> storeAuthInfo(int keyId, String key, int userId) async {
    String authKey = '$keyId:$key';

    await Hive.box('authBox').put('authKey', authKey);
    await Hive.box('authBox').put('userId', userId);

    await client.updateStreamingConnectionAuthenticationKey(authKey);
}
```

**Additional question:** Or do I have to pass an `AuthenticationKeyManager` to the `Client` constructor even in this case, because it does some things in the background like updating keys?

**Comment by yahu1031** [comment]
Yes authentication key manager does things behind the scenes, please check it out [here](https://docs.serverpod.dev/concepts/authentication#app-setup).

**Reply by fleeser** [reply]
Read through some of the clients code and found that it is using the KeyManager to automatically send the key when calling an Endpoint. Extended the AuthenticationKeyManager now in my own implementation using Hive. Still, the main issue remains the same (SocketException). 
Added a new comment down here.

**Reply by yahu1031** [reply]
Could you please share the implemented code snippet, so it will be easy to analyze the issue?

**Comment by yahu1031** [comment]
```dart
void main() async {

  ... // rest code //
  // The session manager keeps track of the signed-in state of the user. You
  // can query it to see if the user is currently signed in and get information
  // about the user.
  sessionManager = SessionManager(
    caller: client.modules.auth,
  );
  await sessionManager.initialize();

  runApp(MyApp());
}
```
Check this out, and try following the implementation, and as the session manager handles the modules(auth) you can't handle it without it.

**Reply by fleeser** [reply]
The documentation says 

> It's most common to use the auth module for handling sign ins, but you can also build your own. If you are building custom authentication methods, use the session.auth.signInUser() method to sign in a user and session.auth.signOutUser() to sign someone out.

In my understanding then, I do not have to include the auth module at all. Or am I missing some fundamental understanding?

Edit: Also the SessionManager stores data I do not want to store like UserInfo and makes use of SharedPreferences.

**Reply by yahu1031** [reply]
Hmmm, then maybe as of my understanding then we must have a custom module to access it from the client. If not, we must have a custom endpoint with a callback URL to your domain.

**Comment by fleeser** [comment, marked as answer]
The problem was that I forgot to start the server with `dart bin/main.dart` ... No comment on that

But even then there were some things clarified:

1. `KeyManager` should be passed when using another storage than the default implementation with SharedPreferences. therefore you should create your own class and extend `AuthenticationKeyManager`
2. `SessionManager` does not have to be passed when you implemented your own logic for auth and key management. For example I wrote a Repository that makes a call to my own endpoint and then managed the results like keys and user information by myself.

Therefore if you created your very own auth endpoint logic, you do not need any package / module.

Thanks to @yahu1031 for fast help.

**Reply by yahu1031** [reply]
Busted me 😭. Good to know things.

**Comment by mozexty** [comment]
Thanks Florian Leeser for sharing this information.

I had tried and failed to build my own auth system. I needed phone number auth.

I'll follow this thread and try again.

Thanks @yahu1031


________________________________

**Reply by vlidholt** [reply]
Phone number authentication would be a great addition to Serverpod. A good starting point would to have a look at how the email authentication is done.
